File: BEMLIB_22.10/01_num_meth/12_pde_poisson/chimney_s0.m
clear all
close all

%===============
% chimney_s
%
% Steady-state temperature distribution
% over the cross-section
% of a square chimney wall
% with Dirichlet BC
%===============

%--------------------------
% prameters and conditions:
%--------------------------

T1=300;         % temperature of inside wall in deg C
T2=10;          % temperature of outside wall in deg C
a=0.5;          % thickness of chimney wall in meters
k = 0.15;       % conductivity
tolerance = 0.00001;

N=32;   % number of nodes in x-direction
maxiter=3000;% maximum # of iterations

%--------
% prepare
%--------


M=N/2;  %  number of nodes in y-direction
b=2*a;

dx=b/N;
dy=a/M;

beta=(dx/dy)^2;
beta1=2.0*(1.0+beta);

%--------------
% initial guess
%--------------

for i=1:N+1
   for j=1:M+1
      T(i,j)=T2;
   end
end

%--------------------
% boundary conditions
%--------------------

for j=1:M+1
   T(1,j)=T2;          % Dirichlet on left edge (outer wall),
   T(N+2,j)=T(N,j);     % and right edge (x=b; also using symmetry)
end

for i=1:M+1
   T(i,1)=T(M+1,M+2-i); % lower edge (from 0 to a; using symmetry)
end

for i=1:N+1     % Dirichlet b.c. and on upper edge (outer wall)
   T(i,M+1)=T2;
end

for i=M+1:N+1   % ...and also on lower edge (bordering inner
   T(i,1)=T1;   % wall; from a to b)
end

%-----------
% iterations
%-----------

for n=1:maxiter         % iterate until convergence

   correction = 0.0;

   for i=2:N+1  % central finite-diff discretization
      for j=2:M         % del^2(T)=0 combined w/ point-Gauss-Siedel
        Told = T(i,j);
        T(i,j)=(T(i+1,j)+T(i-1,j)+beta*(T(i,j+1)+T(i,j-1)))/beta1;
        diff = abs(T(i,j)-Told);
          if (diff>correction)
           correction = diff;
         end
      end
   end

   for i=1:M+1          % reset the bottom (from 0 to a)
      T(i,1)=T(N/2+1,M+2-i);
   end
   for j=1:M+1                  % reset the right edge
      T(N+2,j)=T(N,j);
   end

  correction
  if(correction<tolerance) break; end;

end

%------------------
% end of iterations
%------------------

for i=1:N+1             % set up plotting vectors
  X(i)=dx*(i-1);
end
for j=1:M+1
  Y(j)=dy*(j-1);
end

for i=1:N+1
   X1(i)=4*a-X(i);
end

%---------
% plotting
%---------

hold on

mesh(X,Y,T(1:N+1,:)')   % plotting, labelling, and formatting
mesh(X1,Y,T(1:N+1,:)')
mesh(Y+3.0*a,X-3*a,T(1:N+1,:))
mesh(Y+3.0*a,X1-3*a,T(1:N+1,:))
mesh(X,-2*a-Y,T(1:N+1,:)')
mesh(X1,-2*a-Y,T(1:N+1,:)')
mesh(a-Y,X-3*a,T(1:N+1,:))
mesh(a-Y,X1-3*a,T(1:N+1,:))

xlabel('x (m)','fontsize',15)
ylabel('y (m)','fontsize',15)
zlabel('T (C)','fontsize',15)
set(gca,'fontsize',15)

%title('Temperature over the cross-section of chimney wall')
%axis([0 2.0 0 0.0 0 2.0])
view(-80,40)

%------------------------------------
% compute the flux on the outer wall
%------------------------------------

%---
% top segment:
%---

for i=1:N+1
 flux_top(i) = -k*(T(i,M+1)-T(i,M))/dy;
end

%------------------------------------
% build the flux around the outer wall
%------------------------------------

Ic = 1;  % counter

fluxo(1)=flux_top(1);
perimo(1)=0;
xo(1)=0;
yo(1)=a;

for i=2:N+1
 Ic=Ic+1;
 fluxo(Ic)=flux_top(i);
 perimo(Ic)=perimo(Ic-1)+dx;
 xo(Ic)=xo(Ic-1)+dx;
 yo(Ic)=yo(Ic-1);
end
for i=2:N+1
 Ic=Ic+1;
 fluxo(Ic)=flux_top(N+2-i);
 perimo(Ic)=perimo(Ic-1)+dx;
 xo(Ic)=xo(Ic-1)+dx;
 yo(Ic)=yo(Ic-1);
end
for j=2:N+1
 Ic=Ic+1;
 fluxo(Ic)=flux_top(j);
 perimo(Ic)=perimo(Ic-1)+dy;
 xo(Ic)=xo(Ic-1);
 yo(Ic)=yo(Ic-1)-dy;
end
for j=2:N+1
 Ic=Ic+1;
 fluxo(Ic)=flux_top(N+2-j);
 perimo(Ic)=perimo(Ic-1)+dy;
 xo(Ic)=xo(Ic-1);
 yo(Ic)=yo(Ic-1)-dy;
end
for i=2:N+1
 Ic=Ic+1;
 fluxo(Ic)=flux_top(i);
 perimo(Ic)=perimo(Ic-1)+dx;
 xo(Ic)=xo(Ic-1)-dx;
 yo(Ic)=yo(Ic-1);
end
for i=2:N+1
 Ic=Ic+1;
 fluxo(Ic)=flux_top(N+2-i);
 perimo(Ic)=perimo(Ic-1)+dx;
 xo(Ic)=xo(Ic-1)-dx;
 yo(Ic)=yo(Ic-1);
end
for j=2:N+1
 Ic=Ic+1;
 fluxo(Ic)=flux_top(j);
 perimo(Ic)=perimo(Ic-1)+dy;
 xo(Ic)=xo(Ic-1);
 yo(Ic)=yo(Ic-1)+dy;
end
for j=2:N+1
 Ic=Ic+1;
 fluxo(Ic)=flux_top(N+2-j);
 perimo(Ic)=perimo(Ic-1)+dy;
 xo(Ic)=xo(Ic-1);
 yo(Ic)=yo(Ic-1)+dy;
end

%-----------------------------------
% compute the flux on the inner wall
%-----------------------------------

%------------
% top segment 
%------------

for i=1:M+1
 flux_top(i) = -k*(T(M+i,2)-T(M+i,1))/dy;
end

%------------------------------------
% build the flux around the inner wall
%------------------------------------

Ic = 1;  % counter
fluxi(1)=flux_top(1);
perimi(1)=0;
xi(1)=a;
yi(1)=0;

for i=2:M+1
 Ic=Ic+1;
 perimi(Ic)=perimi(Ic-1)+dx;
 fluxi(Ic)=flux_top(i);
 xi(Ic)=xi(Ic-1)+dx;
 yi(Ic)=yi(Ic-1);
end
for i=2:M+1
 Ic=Ic+1;
 perimi(Ic)=perimi(Ic-1)+dx;
 fluxi(Ic)=flux_top(M+2-i);
 xi(Ic)=xi(Ic-1)+dx;
 yi(Ic)=yi(Ic-1);
end
for j=2:M+1
 Ic=Ic+1;
 perimi(Ic)=perimi(Ic-1)+dy;
 fluxi(Ic)=flux_top(j);
 xi(Ic)=xi(Ic-1);
 yi(Ic)=yi(Ic-1)-dy;
end
for j=2:M+1
 Ic=Ic+1;
 perimi(Ic)=perimi(Ic-1)+dy;
 fluxi(Ic)=flux_top(M+2-j);
 xi(Ic)=xi(Ic-1);
 yi(Ic)=yi(Ic-1)-dy;
end
for i=2:M+1
 Ic=Ic+1;
 fluxi(Ic)=flux_top(i);
 perimi(Ic)=perimi(Ic-1)+dx;
 xi(Ic)=xi(Ic-1)-dx;
 yi(Ic)=yi(Ic-1);
end
for i=2:M+1
 Ic=Ic+1;
 perimi(Ic)=perimi(Ic-1)+dx;
 fluxi(Ic)=flux_top(M+2-i);
 xi(Ic)=xi(Ic-1)-dx;
 yi(Ic)=yi(Ic-1);
end
for j=2:M+1
 Ic=Ic+1;
 fluxi(Ic)=flux_top(j);
 perimi(Ic)=perimi(Ic-1)+dy;
 xi(Ic)=xi(Ic-1);
 yi(Ic)=yi(Ic-1)+dy;
end
for j=2:M+1
 Ic=Ic+1;
 fluxi(Ic)=flux_top(M+2-j);
 perimi(Ic)=perimi(Ic-1)+dy;
 xi(Ic)=xi(Ic-1);
 yi(Ic)=yi(Ic-1)+dy;
end

figure
hold on
%title('Wall flux')
%plot(perimo,fluxo,perimi,fluxi,'--');
%legend('outer','inner')
%xlabel('Perimeter (m)','fontsize',15)
%ylabel('flux (Watt)','fontsize',15)
plot3(xo,yo,zeros(size(xo)))
plot3(xo,yo,fluxo)
plot3(xi,yi,zeros(size(xi)),'r')
plot3(xi,yi,fluxi,'r')
set(gca,'fontsize',15)
xlabel('x(m)','fontsize',15)
ylabel('y(m)','fontsize',15)
zlabel('flux (Watt)','fontsize',15)

File: BEMLIB_22.10/01_num_meth/12_pde_poisson/chimney_s0.f
      program chimney0

      Implicit Double Precision (a-h,o-z)

      Dimension x(500,500),y(500,500)
      Dimension T(500,500)

c--------------------------------------------
c Steady-state temperature distribution in a
c brick chimney wall
c
c Will assume: b = 2a, c = a
c
c Discretization: Dx = Dy = a/N
c                 (beta = 1)
c
c Solution by the point Gauss-Siedel method
c
c-------------------------------------------

c---------------
c Default values
c---------------

      a     = 0.35
      To    = 10.0
      Ti    = 300.0
      Niter = 300
      N     = 16

c     write (6,*)
c     write (6,*) " Please enter a"
c     write (6,*) " --------------"
c     read  (5,*) a

c     write (6,*)
c     write (6,*) " Please enter Touter"
c     write (6,*) " -------------------"
c     read  (5,*) To

c     write (6,*)
c     write (6,*) " Please enter Tinner"
c     write (6,*) " ---------------"
c     read  (5,*) Ti

c     write (6,*)
c     write (6,*) " Please enter the discretization level N"
c     write (6,*) " ---------------------------------------"
c     read  (5,*) N

      write (6,*)
      write (6,*) " Please enter the number of iterations"
      write (6,*) " -------------------------------------"
      read  (5,*) Niter

c----------
c constants
c----------

      Null = 0
      Ntwo = 2

c--------
c prepare
c--------

      Dx = a/N
      Dy = a/N

      N1 = N+1

      NN  = 2*N
      NN1 = NN+1

c------------
c grid points
c------------

      Do i=1,NN1
       xi = (i-1.0)*Dx
       Do j=1,N1
        x(i,j) = xi
        y(i,j) = (j-1.0)*Dy
       End Do
      End Do

c-----------
c initialize
c-----------
   
      Do i=1,NN1
       Do j=1,N1
        T(i,j) = 0.0D0
       End Do
      End Do

c----------
c left wall
c----------

       Do j=1,N1
        T(1,j) = To
       End Do

c-----------
c upper wall
c-----------

       Do i=1,NN1
        T(i,N1) = To
       End Do

c-----------
c lower wall: use the symmetry condition
c-----------

       Do i=1,N1
        T(i,1) = T(N1,N+2-i)
       End Do

       Do i=N1,NN1
        T(i,1) = Ti
       End Do

c-----------
c right wall
c-----------

       Do j=1,N1
        T(NN+2,j) = T(NN,j)
       End Do

c-----------
c Iterations
c-----------

       Icount = 1
       Jcount = 1

  97   Continue

       Do iter=1,Niter

        Do i=2,NN1
         Do j=2,N
          T(i,j) = 0.25*(T(i+1,j)+T(i-1,j)+T(i,j+1)+T(i,j-1))
         End Do
        End Do

c---
c reset the bottom
c---

       Do i=1,N1
        T(i,1) = T(N1,N+2-i)
       End Do

c---
c reset the right 
c---

       Do j=1,N1
        T(NN+2,j) = T(NN,j)
       End Do

       write (6,102) Jcount,T(2,2)

       Icount = Icount+1
       Jcount = Jcount+1

       End Do

c------------------------
c end of batch iterations
c------------------------

       write (6,*)
       write (6,*) " Continue the iterations ?"
       write (6,*) " 1 for YES, 0 for NO"
       read  (5,*) Icon

       If(Icon.ne.0) Go to 97

c----------
c print out
c----------

       open (9,file="chimney_s0.out")

       Do i=1,NN1
        write (9,*) Ntwo
        write (9,100) x(i,1),y(i,1),Null
        write (9,100) x(i,N1),y(i,N1),Null
c       write (6,*) Ntwo
c       write (6,100) x(i,1),y(i,1),Null
c       write (6,100) x(i,N1),y(i,N1),Null
       End Do

       Do j=1,N1
        write (9,*) Ntwo
        write (9,100) x(1,j),y(1,j),Null
        write (9,100) x(NN1,j),y(NN1,j),Null
c       write (6,*) Ntwo
c       write (6,100) x(1,j),y(1,j),Null
c       write (6,100) x(NN1,j),y(NN1,j),Null
       End Do

       Do i=1,NN1
        write (9,*) N1
        Do j=1,N1
c        write (9,100) x(i,j),y(i,j),T(i,j)/300.0
         write (9,100) x(i,j),y(i,j),T(i,j)
        End Do
       End Do

       Do j=1,N1
        write (9,*) NN1
        Do i=1,NN1
c        write (9,100) x(i,j),y(i,j),T(i,j)/300.0
         write (9,100) x(i,j),y(i,j),T(i,j)
        End Do
       End Do

       write (9,*) Null
       close (9)

c-----
c Done
c-----

  100  Format (3(1x,f10.5))
  102  Format (1x,i4,3(1x,f12.8))

       Stop
       End
